### PR TITLE
Processes for BUSCO and minimap2 QC checks

### PR DESCRIPTION
**Pull request by elizabethmcd**
This pull request adds capability for performing QC checks of the assembly with BUSCO (completion estimates) and minimap2 (map reads back to assembly for downstream coverage/annotation checks). These processes were tested with a small nanopore dataset to make sure each process performed as intended (because the hifi test dataset isn't finished yet), but the updated test dataset will be subsampled hifi reads from a tick assembly. Therefore the test profile also includes a parameter for testing BUSCO with arachnid marker genes. This pull request also includes increasing CPU/memory requirements for assembly after talking with Peter about resource requirements.

Pipeline run with `nextflow run main.nf --input test_sheet.csv --outdir test -profile test,docker` because the correct intended test data isn't ready yet for merging into the test-datasets repo, but the workflow works as intended with the test dataset for the metagenomics workflow.

```
[21/3c39f8] process > ARCADIASCIENCE_HIFI2GENOME:HIFI2GENOME:INPUT_CHECK:CHECK_SAMPLESHEET (test_sheet.csv) [100%] 1 of 1, cached: 1 ✔
[da/7866bb] process > ARCADIASCIENCE_HIFI2GENOME:HIFI2GENOME:FLYE (eligo)                                   [100%] 1 of 1, cached: 1 ✔
[84/5b2a72] process > ARCADIASCIENCE_HIFI2GENOME:HIFI2GENOME:BUSCO (eligo)                                  [100%] 1 of 1 ✔
[28/3126f4] process > ARCADIASCIENCE_HIFI2GENOME:HIFI2GENOME:MINIMAP2_SUBWORKFLOW:MINIMAP2_INDEX (1)        [100%] 1 of 1, cached: 1 ✔
[24/6a4c06] process > ARCADIASCIENCE_HIFI2GENOME:HIFI2GENOME:MINIMAP2_SUBWORKFLOW:MINIMAP2_ALIGN (eligo)    [100%] 1 of 1, cached: 1 ✔
[bb/f2e6c5] process > ARCADIASCIENCE_HIFI2GENOME:HIFI2GENOME:CUSTOM_DUMPSOFTWAREVERSIONS (1)                [100%] 1 of 1 ✔
-[Arcadia-Science/hifi2genome] Pipeline completed successfully-
Completed at: 24-Jan-2023 13:49:31
Duration    : 22m 15s
CPU hours   : 0.8 (6.3% cached)
Succeeded   : 2
Cached      : 4
```

### CODE REVIEW

File: modules/nf-core/minimap2/align/main.nf [hunk 1]
```diff
@@ -10,7 +10,7 @@ process MINIMAP2_ALIGN {
 
     input:
     tuple val(meta), path(reads)
-    path reference
```
Comment (mertcelebi) [inline modules/nf-core/minimap2/align/main.nf:13]: Let's move these to `nf-core-modified` and add a comment about why you had to make these changes.

File: modules/nf-core/minimap2/align/main.nf [hunk 2]
```diff
@@ -33,8 +33,8 @@ process MINIMAP2_ALIGN {
     minimap2 \\
         $args \\
         -t $task.cpus \\
-        "${reference ?: reads}" \\
-        "$reads" \\
```
Comment (mertcelebi) [inline modules/nf-core/minimap2/align/main.nf:37]: hmm why this change?

File: modules/nf-core/minimap2/align/main.nf [hunk 3]
```diff
@@ -33,8 +33,8 @@ process MINIMAP2_ALIGN {
     minimap2 \\
         $args \\
         -t $task.cpus \\
-        "${reference ?: reads}" \\
```
Comment (mertcelebi) [inline modules/nf-core/minimap2/align/main.nf:36]: what did syntax do?

File: nextflow.config [hunk 4]
```diff
@@ -10,6 +10,11 @@
 params {
     // Input options
     input                      = null
+
+    // BUSCO options
+    lineage                    = 'auto'
```
Comment (mertcelebi) [inline nextflow.config:15]: do we need to add these to the `nextflow_schama`?

File: workflows/hifi2genome.nf [hunk 5]
```diff
@@ -47,15 +50,32 @@ workflow HIFI2GENOME {
...
+    MINIMAP2_SUBWORKFLOW (
+        FLYE.out.fasta,
+        ch_data.reads
+    )
+    ch_versions = ch_versions.mix(MINIMAP2_SUBWORKFLOW.out.versions)
+
+
+    // dump software versions
+    CUSTOM_DUMPSOFTWAREVERSIONS (
+    ch_versions.unique().collectFile(name: 'collated_versions.yml')
```
Comment (mertcelebi) [inline workflows/hifi2genome.nf]: missing 4 spaces

File: subworkflows/local/minimap2_subworkflow.nf [hunk 6]
```diff
@@ -0,0 +1,30 @@
+//
+// Minimap2 subworkflow to index an assembly and map reads to the assembly
+//
+
+include { MINIMAP2_INDEX } from '../../modules/nf-core/minimap2/index'
+include { MINIMAP2_ALIGN } from '../../modules/nf-core/minimap2/align'
+
+workflow MINIMAP2_SUBWORKFLOW {
+    take:
+    assembly
```
Comment (mertcelebi) [inline subworkflows/local/minimap2_subworkflow.nf]: you should add a comment that this is actually a tuple of meta and values otherwise it's confusing.

File: subworkflows/local/minimap2_subworkflow.nf [hunk 7]
```diff
@@ -0,0 +1,30 @@
+//
+// Minimap2 subworkflow to index an assembly and map reads to the assembly
```
Comment (mertcelebi) [inline subworkflows/local/minimap2_subworkflow.nf:2]: Probably no need to make this a subworkflow functionally at this time, but it's cool

File: subworkflows/local/minimap2_subworkflow.nf [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    main:
+    ch_versions = Channel.empty()
+
+    // build index
+    MINIMAP2_INDEX(assembly)
+    ch_versions = ch_versions.mix(MINIMAP2_INDEX.out.versions)
+    ch_index = MINIMAP2_INDEX.out.index
+
+    // align reads to index
+    MINIMAP2_ALIGN(reads, ch_index, "true", "false", "true")
```
Comment (mertcelebi) [inline subworkflows/local/minimap2_subworkflow.nf]: Two questions:
* Do we want to parameterize the "true", "false" stuff?
* Are you sure this is the expected usage of the `MINIMAP2_ALIGN`? The strings are throwing me off as I don't know how Groovy passes them on to the underlying process. But if it works like any other language and passes them as a string (rather than a boolean) "false" is still truthy, which is not what you want.

File: bin/check_samplesheet.py [hunk 9]
```diff
@@ -79,9 +79,9 @@ def check_samplesheet(file_in, file_out):
                 if reads:
                     if reads.find(" ") != -1:
                         print_error("Fastq file contains spaces!", "Line", line)
-                    if not reads.endswith(".fastq.gz") and not reads.endswith(".fq.gz"):
+                    if not reads.endswith(".fastq.gz") and not reads.endswith(".fq.gz") and not reads.endswith(".fastq") and not reads.endswith(".fq"):
```
Comment (mertcelebi) [inline bin/check_samplesheet.py:82]: hmm, does the underlying modules work with both compressed and regular files?

**Review by mertcelebi — COMMENTED** [review]
Adding some comments and questions for now

File: subworkflows/local/minimap2_subworkflow.nf (hunk 8 above)
Comment (elizabethmcd) [reply subworkflows/local/minimap2_subworkflow.nf]: I have further dug into how this is actually working in https://github.com/Arcadia-Science/metagenomics/pull/26 and it's not quite what the above is doing I've discovered. I will need to change this because in this case I do actually want the BAM format and not PAF, whereas I had to take out this behavior entirely in the metagenomics workflow.

File: modules/nf-core/minimap2/align/main.nf (hunk 2 above)
Comment (elizabethmcd) [reply modules/nf-core/minimap2/align/main.nf:37]: I thought that there was an issue with this before, not sure if that's still the case but is works fine with just $reference or $index in this case because I already perform the indexing as part of the prior indexing minimap2 module

File: modules/nf-core/minimap2/align/main.nf (hunk 3 above)
Comment (elizabethmcd) [reply modules/nf-core/minimap2/align/main.nf:36]: I don't know what it was doing before and it seemed to be causing problems and the modified change works

File: nextflow.config (hunk 4 above)
Comment (elizabethmcd) [reply nextflow.config:15]: yes, I need to do that

File: subworkflows/local/minimap2_subworkflow.nf (hunk 6 above)
Comment (elizabethmcd) [reply subworkflows/local/minimap2_subworkflow.nf]: addressed

File: workflows/hifi2genome.nf (hunk 5 above)
Comment (elizabethmcd) [reply workflows/hifi2genome.nf]: I don't understand this comment?

File: subworkflows/local/minimap2_subworkflow.nf (hunk 7 above)
Comment (elizabethmcd) [reply subworkflows/local/minimap2_subworkflow.nf:2]: yeah for things that go together like indexing a reference and then mapping a reference I like to put it in a dedicated subworkflow just so that it's more clear. It follows how I've done the mapping_depth subworkflow for the illumina workflow and this minimap2 subworkflow will also have an added depth calculation step next, so this structure was to keep them consistent

File: bin/check_samplesheet.py (hunk 9 above)
Comment (elizabethmcd) [reply bin/check_samplesheet.py:82]: yes

Comment (elizabethmcd) [reply bin/check_samplesheet.py:82]: I don't have an explicit test for this however, but I have tested it separately with some of Peter's data that he didn't compress

File: nextflow.config (hunk 4 above)
Comment (elizabethmcd) [reply nextflow.config:15]: addressed

**Comment by elizabethmcd** [comment]
This now works with a proper hifi dataset of two Ecoli strains of subsetted reads (described in PR https://github.com/Arcadia-Science/test-datasets/pull/11) and the workflow completes: 
```
[b8/d0d867] process > ARCADIASCIENCE_HIFI2GENOME:HIFI2GENOME:INPUT_CHECK:CHECK_SAMPLESHEET (test.csv)        [100%] 1 of 1 ✔
[3d/6ef331] process > ARCADIASCIENCE_HIFI2GENOME:HIFI2GENOME:FLYE (ecoli_H10407)                             [100%] 2 of 2 ✔
[e9/a92c0c] process > ARCADIASCIENCE_HIFI2GENOME:HIFI2GENOME:BUSCO (ecoli_H10407)                            [100%] 2 of 2 ✔
[f1/12e4c9] process > ARCADIASCIENCE_HIFI2GENOME:HIFI2GENOME:MINIMAP2_SUBWORKFLOW:MINIMAP2_INDEX (2)         [100%] 2 of 2 ✔
[d9/9e42e6] process > ARCADIASCIENCE_HIFI2GENOME:HIFI2GENOME:MINIMAP2_SUBWORKFLOW:MINIMAP2_ALIGN (ecoli_k12) [100%] 2 of 2 ✔
[06/1e0fa4] process > ARCADIASCIENCE_HIFI2GENOME:HIFI2GENOME:CUSTOM_DUMPSOFTWAREVERSIONS (1)                 [100%] 1 of 1 ✔
-[Arcadia-Science/hifi2genome] Pipeline completed successfully-
Completed at: 27-Jan-2023 13:11:48
Duration    : 3m 41s
CPU hours   : 0.2
Succeeded   : 10
```
This changes the test config to have lineage enterobacterales to fit with ecoli instead of using tick reads as a test because they are too big and complex.

File: nextflow_schema.json [hunk 10]
```diff
@@ -13,19 +13,35 @@
             "properties": {
                 "input": {
                     "type": "string",
-                    "default": null,
+                    "default": "None",
```
Comment (mertcelebi) [inline nextflow_schema.json]: Null is probably better as it'd be easier to catch for presence or not later in the pipeline

File: nextflow_schema.json [hunk 11]
```diff
@@ -13,19 +13,35 @@
...
-                    "default": null
+                    "default": "None"
                 }
             },
             "required": [
                 "input",
                 "outdir"
             ]
         },
+        "busco_options": {
```
Comment (mertcelebi) [inline nextflow_schema.json:29]: 👍

File: workflows/hifi2genome.nf (hunk 5 above)
Comment (mertcelebi) [reply workflows/hifi2genome.nf]: You should indent it 4 spaces as are the previous multi-line process calls

**Review by mertcelebi — APPROVED** [review]
Approving with a small comment. Please address those and merge away.

File: nextflow_schema.json (hunk 10 above)
Comment (elizabethmcd) [reply nextflow_schema.json]: addressed
